homeassistant/components/keyboard_remote/__init__.py

```diff
@@ -36,177 +42,271 @@
     config = config.get(DOMAIN)
 
-    keyboard_remote = KeyboardRemote(hass, config)
-
-    def _start_keyboard_remote(_event):
-        keyboard_remote.run()
-
-    def _stop_keyboard_remote(_event):
-        keyboard_remote.stop()
-
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, _start_keyboard_remote)
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, _stop_keyboard_remote)
+    remote = KeyboardRemote(hass, config)
+    await remote.async_setup()
 
     return True
 
 
-class KeyboardRemoteThread(threading.Thread):
-    """This interfaces with the inputdevice using evdev."""
+class KeyboardRemote:
+    """Manage device connection/disconnection using inotify to asynchronously monitor."""
 
-    def __init__(self, hass, device_name, device_descriptor, key_value):
-        """Construct a thread listening for events on one device."""
+    def __init__(self, hass, config):
+        """Create handlers and setup dictionaries to keep track of them."""
         self.hass = hass
-        self.device_name = device_name
-        self.device_descriptor = device_descriptor
-        self.key_value = key_value
-
-        if self.device_descriptor:
-            self.device_id = self.device_descriptor
-        else:
-            self.device_id = self.device_name
-
-        self.dev = self._get_keyboard_device()
-        if self.dev is not None:
-            _LOGGER.debug("Keyboard connected, %s", self.device_id)
-        else:
-            _LOGGER.debug(
-                "Keyboard not connected, %s. " "Check /dev/input/event* permissions",
-                self.device_id,
-            )
-
-            id_folder = "/dev/input/by-id/"
-
-            if os.path.isdir(id_folder):
-                from evdev import InputDevice, list_devices
-
-                device_names = [
-                    InputDevice(file_name).name for file_name in list_devices()
-                ]
-                _LOGGER.debug(
-                    "Possible device names are: %s. "
-                    "Possible device descriptors are %s: %s",
-                    device_names,
-                    id_folder,
-                    os.listdir(id_folder),
-                )
+        self.handlers_by_name = {}
+        self.handlers_by_descriptor = {}
+        self.active_handlers_by_descriptor = {}
+        self.monitor_task = None
 
-        threading.Thread.__init__(self)
-        self.stopped = threading.Event()
-        self.hass = hass
+        for dev_block in config:
+            handler = self.DeviceHandler(hass, dev_block)
+            descriptor = dev_block.get(DEVICE_DESCRIPTOR)
+            if descriptor is not None:
+                self.handlers_by_descriptor[descriptor] = handler
+            else:
+                name = dev_block.get(DEVICE_NAME)
+                self.handlers_by_name[name] = handler
 
-    def _get_keyboard_device(self):
-        """Get the keyboard device."""
-        from evdev import InputDevice, list_devices
+    async def async_setup(self):
```

Comment:
This doesn't need to be a coroutine, since we don't await inside. Callbacks are a bit faster than coroutines, so we try to be strict about this selection.
```suggestion
    def async_setup(self):
```